session.go

```diff
@@ -338,35 +338,35 @@ func sqlForLog(sql string) string {
 	return sql
 }
 
+var sysSessionPool = sync.Pool{}
+
 // ExecRestrictedSQL implements RestrictedSQLExecutor interface.
 // This is used for executing some restricted sql statements, usually executed during a normal statement execution.
 // Unlike normal Exec, it doesn't reset statement status, doesn't commit or rollback the current transaction
 // and doesn't write binlog.
 func (s *session) ExecRestrictedSQL(ctx context.Context, sql string) (ast.RecordSet, error) {
-	s.prepareTxnCtx()
-	charset, collation := s.sessionVars.GetCharsetInfo()
-	rawStmts, err := s.ParseSQL(sql, charset, collation)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	if len(rawStmts) != 1 {
-		log.Errorf("ExecRestrictedSQL only executes one statement. Too many/few statement in %s", sql)
-		return nil, errors.New("wrong number of statement")
+	// Use special session to execute the sql.
+	var se *session
+	tmp := sysSessionPool.Get()
+	if tmp != nil {
+		se = tmp.(*session)
+	} else {
+		var err error
+		se, err = createSession(s.store)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		varsutil.SetSessionSystemVar(se.sessionVars, variable.AutocommitVar, types.NewStringDatum("1"))
+		se.sessionVars.CommonGlobalLoaded = true
+		se.sessionVars.InRestrictedSQL = true
 	}
-	// Some execution is done in compile stage, so we reset it before compile.
-	st, err := Compile(s, rawStmts[0])
-	if err != nil {
-		log.Errorf("Compile %s with error: %v", sql, err)
-		return nil, errors.Trace(err)
+	defer sysSessionPool.Put(se)
```

Comment:
The returned `ast.RecordSet` still access session variables, it's not safe to just put it in the pool.
We should get all rows, then return rows instead of `ast.RecordSet`.